### PR TITLE
Improved IDE experience for JUnit5 - visual comparison failure

### LINKED ISSUE #1663: Improve IDE UX comparison failure for JUnit5
When JUnit4 is used and the test is invoked from IDE (IDEA or Eclipse), and we are getting "arguments are different" MockitoVerificationError, the IDE gets nice "click to see the difference" link in the test failure, providing neat comparison pop-up.

We want to provide this feature for JUnit5, too. We do not want to add a new dependency on mockito (e.g. OpenTest4J). Rather, we can use reflection just like we do it now for JUnit4 or how AssertJ integrates with JUnit5.

Spin-off from #1656

### PR DESCRIPTION
**Pull request by kriegfrj**
When JUnit5+Mockito is used in modern IDE (IDEA, Eclipse) we now show "visual" comparison failure pop-up for certain Mockito exceptions (such as ArgumentsAreDifferent).

Fixes #1663.

### CODE REVIEW

File: src/main/java/org/mockito/internal/junit/ExceptionFactory.java [hunk 1]
```diff
@@ -4,49 +4,63 @@
...
+                    @Override
+                    public AssertionError create(String message, String wanted, String actual) {
+                        return new org.mockito.exceptions.verification.junit.ArgumentsAreDifferent(message, wanted, actual);
+                    }
+                };
+            } catch (Throwable onlyIfJUnit3IsNotAvailable) {
+            }
         }
-        return true;
+        // The following is neater but requires -source and -target >= 1.8
```
Comment (TimvdLippe) [inline src/main/java/org/mockito/internal/junit/ExceptionFactory.java]: We will get to that when we switch to Java 1.8. Let's remove these comments for now.

**Comment by kriegfrj** [comment]
> Do you have any instructions on how I can test this PR to verify the difference in IDE behavior?

* Create a simple JUnit 5 test that fails with `ArgumentsAreDifferent`:
````java
package mytest;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

import org.junit.jupiter.api.Test;

public class Test {
    @SuppressWarnings("unchecked")
    @Test
    public void test() {
        Comparable<Object> c = (Comparable<Object>)mock(Comparable.class);
        c.compareTo("test");
        
        verify(c).compareTo("somethingElse");
    }
}
````
* Make sure that the snapshot version of Mockito is on your classpath.
* Make sure JUnit 5 is on the test classpath. If you're using Mockito's workspace, you can do this with the following dependency declarations:
    * `testCompile libraries.junitJupiterApi`
    * `testRuntime libraries.junitJupiterEngine`
* Run the test in the IDE using the JUnit 5 runner.
* In Eclipse, you should get the following:
ArgumentsAreDifferent
I'm not sure how it looks in IDEA or Android Studio, but I'm sure it's similar.

File: src/main/java/org/mockito/internal/junit/ExceptionFactory.java (hunk 1 above)
Comment (kriegfrj) [reply src/main/java/org/mockito/internal/junit/ExceptionFactory.java]: Done.

**Comment by kriegfrj** [comment]
This implementation is not perfect. Under Eclipse, if you're using the JUnit 4 but happen to have OpenTest on the classpath, then you'll use the visual diff capability again. This is because the JUnit 4 runner in Eclipse doesn't try and catch/handle `org.opentest4j.AssertionFailedError`. I'm not sure of the behaviour in IDEA.

The only way around that would be for `ExceptionFactory.createArgumentsAreDifferent()` to examine the caller stack to see if it has been invoked from a JUnit 4 or a JUnit 5 context, and throw the most appropriate exception. This would complicate the implementation and make it run slower too. It would be better, IMO, if the JUnit 4 runner in the IDEs were made aware of the OpenTest4J exception and could handle it too.

**Comment by mockitoguy** [comment]
> This implementation is not perfect. Under Eclipse, if you're using the JUnit 4 but happen to have OpenTest on the classpath, then you'll use the visual diff capability again.

Why do you call this behavior imperfect?

**Comment by kriegfrj** [comment]
> > This implementation is not perfect. Under Eclipse, if you're using the JUnit 4 but happen to have OpenTest on the classpath, then you'll use the visual diff capability again.
>
> Why do you call this behavior imperfect?
Sorry, it was a typo. It should have read "... then you'll *lose* the visual diff capability again." Perfect behaviour, in my mind, would be that it determined which runner was actually running (as opposed to merely being on the classpath), and throw the exception that was appropriate for that runner. But I think this is an edge case that probably isn't worth the extra complexity. Not perfect, but good enough.

File: src/test/java/org/mockito/internal/verification/VerificationOverTimeImplTest.java [hunk 2]
```diff
@@ -10,7 +10,7 @@
 import org.junit.rules.ExpectedException;
 import org.mockito.Mock;
 import org.mockito.exceptions.base.MockitoAssertionError;
-import org.mockito.exceptions.verification.junit.ArgumentsAreDifferent;
+import org.mockito.exceptions.verification.opentest4j.ArgumentsAreDifferent;
```
[the remark this answers is not in the document]
Comment (kriegfrj) [reply src/test/java/org/mockito/internal/verification/VerificationOverTimeImplTest.java]: Yes. This was the "imperfect" behaviour I was referring to. Even though it's only using the JUnit 4 runner, it finds OpenTest4J on the classpath and uses it.

File: src/main/java/org/mockito/exceptions/verification/junit/package-info.java [hunk 3]
```diff
@@ -4,6 +4,8 @@
  */
 
 /**
- * JUnit integration to provide better support for junit runners in IDEs.
+ * JUnit 3 integration to provide better support for junit 3 runners in IDEs.
```
[the remark this answers is not in the document]
Comment (kriegfrj) [reply src/main/java/org/mockito/exceptions/verification/junit/package-info.java]: Will do. This is a hangover from the earlier prototype which also supported `org.junit.ComparisonFailure` as well as `junit.framework.ComparisonFailure`.

File: src/main/java/org/mockito/internal/junit/ExceptionFactory.java [hunk 4]
```diff
@@ -4,49 +4,57 @@
...
 
     /**
-     * Don't inline this class! It allows create the JUnit-ArgumentsAreDifferent exception without the need to use reflection.
-     * <p>
-     * If JUnit is not available a call to {@link #create(String, String, String)} will throw a {@link NoClassDefFoundError}.
-     * The {@link NoClassDefFoundError} will be thrown by the class loader cause the JUnit class {@link ComparisonFailure}
-     * can't be loaded which is a upper class of ArgumentsAreDifferent.
+     * Returns an AssertionError that describes the fact that the arguments of an invocation are different.
+     * If OpenTest is on the class path, it returns a class that extends from {@link org.opentest4j.AssertionFailedError},
+     * or else if JUnit 3 is on the class path it returns a class that extends from JUnit 3's
```
[the remark this answers is not in the document]
Comment (kriegfrj) [reply src/main/java/org/mockito/internal/junit/ExceptionFactory.java]: As above.

**Comment by kriegfrj** [comment]
> This looks great, thank you!!!
> 
> Can you do 2 more checks for us?
> 
> * does it work with IDEA, too?

This will be difficult for me to do, as I don't have IDEA installed. But I can't see why it wouldn't work

> * can you build the project and inspect all pom files (they should be somewhere under "build" dirs) and ensure that we don't ship the new dependency?

Will do. I did note that I need to make sure that an optional Import-Package is added to the MANIFEST.MF too.

**Comment by mockitoguy** [comment]
> This will be difficult for me to do, as I don't have IDEA installed.

Grab it from JetBrains site, IDEA community/open source edition is free and easy to install

**Comment by mockitoguy** [comment]
> Sorry, it was a typo.

Got it! Thank you for confirming. Yes, the ideal behavior is that we get clean visual diff for all IDEs. Let's move on. It seems that having OpenTest4J + JUnit4 on classpath is an edge case. Also, I would hope that Eclipse will catch up and start handling OpenTest4J exceptions.

@TimvdLippe, this is an interesting argument for our discussion about adding hard dependency on OpenTest4J. If we add such dependency today, then Eclipse + JUnit4 + Mockito (with OpenTest4J) combo loses visual diff feature (unless we write code to detect the runner - a fragile complexity better avoided). Just an FYI. I don't mind to restart the discussion.

Thanks guys!!!

**Comment by TimvdLippe** [comment]
> If we add such dependency today, then Eclipse + JUnit4 + Mockito (with OpenTest4J) combo loses visual diff feature (unless we write code to detect the runner - a fragile complexity better avoided). 

I am absolutely fine with not adding the hard-dependency :smile:

**Comment by kriegfrj** [comment]
> Got it! Thank you for confirming. Yes, the ideal behavior is that we get clean visual diff for all IDEs. Let's move on.

👍 

> It seems that having OpenTest4J + JUnit4 on classpath is an edge case. Also, I would hope that Eclipse will catch up and start handling OpenTest4J exceptions.
> 

Just to clarify (in case it wasn't clear): The problem is not merely having both OpenTest4J and JUnit4 on the class path. JUnit 5 and OpenTest4J have been deliberately designed so that they can happily co-exist with JUnit 4 on the same classpath.

The problem (in Eclipse at least) is if you mix the JUnit 4 _runner_ with OpenTest4J exceptions (which this current PR will do if you have OpenTest4J on the classpath) - the visual diff will not work, because Eclipse's implementation of the JUnit 4 runner special-cases the JUnit 3 & 4 `ComparisonFailure`, but not the OpenTest4J `AssertionFailedError` (conversely, Eclipse's JUnit 5 runner handles only `AssertionFailedError` and not `ComparisonFailure`). If you use the JUnit 5 runner in Eclipse the visual diff will work, even if JUnit 4 is on the classpath.

> @TimvdLippe, this is an interesting argument for our discussion about adding hard dependency on OpenTest4J. If we add such dependency today, then Eclipse + JUnit4 + Mockito (with OpenTest4J) combo loses visual diff feature (unless we write code to detect the runner - a fragile complexity better avoided). Just an FYI. I don't mind to restart the discussion.

Actually, the issue of losing the visual diff doesn't have anything to do with whether the dependency on OpenTest4J is hard or soft. It has more to do with how the decision is made about which exception to throw. The above problem would still occur even if we made the dependency on OpenTest4J a hard one, unless we did the same stack trace trick (ie, check which runner invoked us and throw the appropriate exception).

> Thanks guys!!!

Welcome! Honoured to be able to meaningfully contribute to such a popular project!

**Comment by kriegfrj** [comment]
IDEA verification:

File: src/main/java/org/mockito/internal/junit/ExceptionFactory.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (kriegfrj) [reply src/main/java/org/mockito/internal/junit/ExceptionFactory.java]: I started to implement this change and had a rethink. Merely saying "...if JUnit is on the class path..." is inaccurate. Specifically, `junit.framework.ComparisonFailure` must be on the classpath, which is included in JUnit 3 & 4 but not 5. So the current implementation will not throw a ComparisonFailure if JUnit 5 is on the classpath but JUnit 3/4 is not.

**Comment by TimvdLippe** [comment]
Could you maybe add a new subproject which has some sample tests with instructions on how to inspect the diff? That would ease future development so that we can check if regressed on that. Thanks for all the hard work, good progress!

**Comment by kriegfrj** [comment]
> > * can you build the project and inspect all pom files (they should be somewhere under "build" dirs) and ensure that we don't ship the new dependency?
> 
> Will do. I did note that I need to make sure that an optional Import-Package is added to the MANIFEST.MF too.

I did this - I can confirm that opentest4j does not appear as a dependency in any of the generated pom files.

I also added the manifest change: `org.opentest4j` is now an optional import to the OSGi manifest so that it would work properly in an OSGi environment.

**Comment by kriegfrj** [comment]
> Could you maybe add a new subproject which has some sample tests with instructions on how to inspect the diff? That would ease future development so that we can check if regressed on that. Thanks for all the hard work, good progress!

I'll sleep on it, but to be honest my initial thoughts are that I feel this is overreaching a little. I think the Mockito tests should focus on the fact that it throws `AssertionFailedError` correctly, which I think they already do. Added to which I'm not sure I have the time.

Perhaps this could be the subject of a separate issue/PR which I can get to if I find time, or else someone else can take it on.

**Comment by mockitoguy** [comment]
> Could you maybe add a new subproject which has some sample tests with instructions on how to inspect the diff? That would ease future development so that we can check if regressed on that. Thanks for all the hard work, good progress!

I suggest we move on. We won't be able to automatically test how IDEs integrate with xUnit runners. Manual tests will be forgotten. We never had tests for this feature and it did not regress. Eventual regression will not be a blocker. If we encounter one, then we can prioritize a solution.

File: src/test/java/org/mockito/internal/verification/VerificationOverTimeImplTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mockitoguy) [reply src/test/java/org/mockito/internal/verification/VerificationOverTimeImplTest.java]: Thanks!

File: src/main/java/org/mockito/internal/junit/ExceptionFactory.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (mockitoguy) [reply src/main/java/org/mockito/internal/junit/ExceptionFactory.java]: Thanks!

File: src/main/java/org/mockito/exceptions/verification/junit/package-info.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mockitoguy) [reply src/main/java/org/mockito/exceptions/verification/junit/package-info.java]: OK

**Comment by TimvdLippe** [comment]
I didn't count it as a blocker, more as a future improvement. I am okay with merging as-is.
